specs/interop/control-flow.md

```diff
@@ -0,0 +1,330 @@
+```
+
+Emitted when a block gets replaced for any reason.
+
+## Supervisor `->` Node
+
+Messages that a node should watch for, originating from supervisor, handled by node. This also includes control signals
+that the supervisor can ask the node to perform. For the used types, refer [this](#types) section.
+
+*Note: Headings represent the actual name of the methods over RPC, payload is the serialized version of the given type.*
+
+`-> indicates return`
+
+### Control Signals
+
+RPC calls that are relevant to managing the node via supervisor. These are directly called by the supervisor.
+
+#### interop_pullEvent
+
+```javascript
+payload: none -> One of the event from node events (previous section).
+```
+
+When websocket is not available, supervisor can use this method to get the next event from node.
+
+#### interop_anchorPoint
+
+```javascript
+payload: none -> DerivedBlockRefPair {
+    Source: L1 BlockRef that rollup starts after (no derived transaction)
+    Derived: L2 BlockRef that rollup starts from (no txn, pre-configuration state)
+}
+```
+
+Returns the genesis block ref for L1 and L2 that the current node used as anchor point.
+
+#### interop_invalidateBlock
+
+```javascript
+payload: BlockSeal //L2's Block
+```
+
+Based on some dependency or L1 changes, supervisor can instruct the L2 to invalidate a specific block.
+
+*(Suggestion: BlockSeal can be replaced with Hash, as only that is being used in op-node.)*
+
+#### interop_provideL1
+
+```javascript
+payload: BlockRef //L1 Block
+```
+
+Supervisor sends the next L1 block to the node. Ideally sent after the node emits `exhausted-l1`.
+
+#### interop_reset
+
+```javascript
+payload: (lUnsafe, xUnsafe, lSafe, xSafe, finalized: BlockID)
+```
+
+Forces a reset to a specific local-unsafe/local-safe/finalized starting point only if the blocks did exist. Resets may
+override local-unsafe, to reset the very end of the chain. Resets may override local-safe, since post-interop we need
+the local-safe block derivation to continue.
+
+### DB
+
+RPC calls that a node should watch for, originating from supervisor that is called on DB updates for relevant block
+safety info for a given chain and block.
+
+#### interop_updateCrossSafe
+
+```javascript
+payload: (derived: BlockID, derivedFrom: BlockID)
+```
+
+Signal that a block can be promoted to cross-safe.
+
+#### interop_updateCrossUnsafe
+
+```javascript
+payload: BlockID
+```
+
+Signal that a block can be promoted to cross-unsafe.
+
+#### interop_updateFinalized
+
+```javascript
+payload: BlockID
+```
+
+Signal that a block can be marked as finalized.
+
+### Sync Methods
+
+RPC methods that are relevant for syncing the supervisor. These are directly called by the supervisor for fetching L2
+data.
+
+#### interop_fetchReceipts
+
+```javascript
+payload: Hash -> Receipts //L2 block hash
+```
+
+Fetches all transaction receipts in a given L2 block.
+
+#### interop_l2BlockRefByTimestamp
+
+```javascript
+payload: uint64 -> L2BlockRef
+```
```

Comment:
We should change this to `BlockRef`. Using the `L2BlockRef` in the API is completely unnecessary, and a mistake in the op-supervisor implementation.